classes/SpecificPrice.php

```diff
@@ -337,24 +383,148 @@ protected static function formatIntInQuery($first_value, $second_value)
+            // So it's ok the set the value to 0 for those fields to improve the cache efficiency
+            // Note that the variableName from the DB needs to match the function args name
+            // I.e. if the computeKey args are converted at some point in camelCase, we will need to introduce a
+            // snakeCase to camelCase conversion of $variableName
+            foreach (array_keys(self::$_no_specific_values) as $variableName) {
+                ${$variableName} = 0;
+            }
+        }
+        return ((int)$id_product.'-'.(int)$id_shop.'-'.(int)$id_currency.'-'.(int)$id_country.'-'.
+            (int)$id_group.'-'.(int)$quantity.'-'.(int)$id_product_attribute.'-'.(int)$id_cart.'-'.
+            (int)$id_customer.'-'.(int)$real_quantity);
+    }
+
+    /**
+     * Returns the specificPrice information related to a given productId and context
+     *
+     * @param int  $id_product
+     * @param int  $id_shop
+     * @param int  $id_currency
+     * @param int  $id_country
+     * @param int  $id_group
+     * @param int  $quantity
+     * @param int  $id_product_attribute
+     * @param int  $id_customer
+     * @param int  $id_cart
+     * @param int  $real_quantity
+     *
+     * @return array
+     */
+    public static function getSpecificPrice($id_product, $id_shop, $id_currency, $id_country, $id_group, $quantity,
+                                            $id_product_attribute = null, $id_customer = 0, $id_cart = 0,
+                                            $real_quantity = 0)
     {
         if (!SpecificPrice::isFeatureActive()) {
             return array();
         }
         /*
-        ** The date is not taken into account for the cache, but this is for the better because it keeps the consistency for the whole script.
-        ** The price must not change between the top and the bottom of the page
+         * The date is not taken into account for the cache, but this is for the better because it keeps the consistency
+         * for the whole script.
+         * The price must not change between the top and the bottom of the page
         */
 
+        if (!self::couldHaveSpecificPrice($id_product)) {
+            return array();
+        }
+
         static $psQtyDiscountOnCombination = null;
         if ($psQtyDiscountOnCombination === null) {
             $psQtyDiscountOnCombination = Configuration::get('PS_QTY_DISCOUNT_ON_COMBINATION');
-        }
-
-        $key = ((int)$id_product.'-'.(int)$id_shop.'-'.(int)$id_currency.'-'.(int)$id_country.'-'.(int)$id_group.'-'.(int)$quantity.'-'.(int)$id_product_attribute.'-'.(int)$id_cart.'-'.(int)$id_customer.'-'.(int)$real_quantity);
-        if (!array_key_exists($key, SpecificPrice::$_specificPriceCache)) {
+            // no need to compute the key the first time the function is called, we know the cache has not
+            // been computed yet
+            $key = null;
+        } else {
+            $key = self::computeKey(
```

Comment:
mmm for me it should remains self. We don't want to encourage overloading of such internal functions (especially if we want to introduce later an API).
I actually think it would be great to declare those internal functions as final